multidownloader/evm_multidownloader_syncers.go

```
@@ -0,0 +1,103 @@
+	dh.log.Debugf("EVMMultidownloader.FilterLogs: received query: %+v", query)
+	defer dh.log.Debugf("EVMMultidownloader.FilterLogs: finished query: %+v", query)
+	logQuery := mdrtypes.NewLogQueryFromEthereumFilter(query)
+	for !dh.IsAvailable(logQuery) {
+		dh.log.Infof("EVMMultidownloader.FilterLogs: waiting %s for logs to be available: %s",
+			dh.cfg.WaitPeriodToCheckCatchUp.String(), logQuery.String())
+		select {
+		case <-time.After(dh.cfg.WaitPeriodToCheckCatchUp.Duration):
+		case <-ctx.Done():
+			return nil, fmt.Errorf("EVMMultidownloader.FilterLogs: "+
+				"context done while waiting for logs %s to be available: %w",
+				logQuery.String(), ctx.Err())
+		}
+	}
+	logs, err := dh.storage.GetEthLogs(nil, logQuery)
+	if err != nil {
+		return nil, fmt.Errorf("EVMMultidownloader.FilterLogs: cannot get logs: %w", err)
+	}
+	dh.log.Debugf("EVMMultidownloader.FilterLogs(%d - %d): len(logs)= %d", query.FromBlock, query.ToBlock, len(logs))
+
+	return logs, nil
+}
+
+// HeaderByNumber gets the block header for the given block number from storage or ethClient
+func (dh *EVMMultidownloader) HeaderByNumber(ctx context.Context, number *big.Int) (*aggkittypes.BlockHeader, error) {
+	dh.log.Debugf("EVMMultidownloader.HeaderByNumber: received number: %s", number.String())
+	defer dh.log.Debugf("EVMMultidownloader.HeaderByNumber: finished number: %s", number.String())
+	if number.Cmp(big.NewInt(0)) < 0 {
+		return nil, fmt.Errorf("EVMMultidownloader.HeaderByNumber: negative block number not supported=%s", number.String())
+	}
+
+	block, _, err := dh.storage.GetBlockHeaderByNumber(nil, number.Uint64())
+	if err != nil {
+		return nil, fmt.Errorf("EVMMultidownloader.HeaderByNumber: cannot get BlockHeader number=%d: %w",
+			number.Uint64(), err)
+	}
+	if block != nil {
+		return block, nil
+	}
+	// This is a fallback mechanism in case the block is not found in storage (must be on storage!)
```

If it must be on storage, should we just return an error in this case, or at least, store this fallback result to `db`?

---

No, it can be that the required block in not in Multidownloader: currently the syncer as asking for empty blocks as synchronization points, these blocks are not synced by Multidownloader. That's the reason that is not an error,  nor even a warning